src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,289 @@
+      reject(err);
+    });
+
+    child.on("exit", (code: number) => {
+      // This shouldn't happen, so ensure that we surface a non-zero exit code.
+      process.exitCode = code === 0 ? 1 : code;
+      reject(
+        new Error(
+          `${START_ERROR}\nThe detached instance exited with error code: ${code}`
+        )
+      );
+    });
+  });
+
+  // destroy the ReadableStream exposed by the child process, to allow the
+  // parent to exit gracefully.
+  child.stderr.destroy();
+  child.unref();
+  child.disconnect();
+
+  const flavor = instanceInfo.flavor;
+  const { host, port } = instanceInfo.server;
+  const cmd =
+    process.platform === "win32"
+      ? path.basename(process.execPath)
+      : [process.execPath, ...process.execArgv, module, ...childArgs].join(" ");
+
+  const pid = child.pid;
+  const startTime = Date.now();
+
+  const instance: DetachedInstance = {
+    startTime,
+    pid,
+    instanceName: createInstanceName(),
+    host,
+    port,
+    flavor,
+    cmd,
+    version
+  };
+
+  while (true) {
+    const instanceFilename = getInstanceFilePath(instance.instanceName);
+    try {
+      await writeFile(instanceFilename, JSON.stringify(instance), {
+        // wx means "Open file for writing, but fail if the path exists". see
+        // https://nodejs.org/api/fs.html#file-system-flags
+        flag: "wx",
+        encoding: FILE_ENCODING
+      });
+      break;
+    } catch (err) {
+      switch ((err as NodeJS.ErrnoException).code) {
+        case "EEXIST":
+          // an instance already exists with this name
+          instance.instanceName = createInstanceName();
+          break;
+        case "ENOENT":
+          // we don't check whether the folder exists before writing, as that's
+          // a very uncommon case. Catching the exception and subsequently
+          // creating the directory is faster in the majority of cases.
+          await mkdir(dataPath, { recursive: true });
+          break;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  return instance;
+}
+
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  let files: string[];
+  let processes: ProcessDescriptor[];
+  try {
+    [files, processes] = await Promise.all([readdir(dataPath), psList()]);
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
+      throw err;
+    }
+    return [];
+  }
+  const instances: DetachedInstance[] = [];
+
+  const loadingInstancesInfos = files.map(async instanceName => {
+    let shouldRemoveFile = false;
+    try {
+      // getDetachedInstanceByName() throws if the instance file is not found or
+      // cannot be parsed
+      const instance = await getDetachedInstanceByName(instanceName);
```

Comment:
Wonder if we need to watch out for invalid files, or if the `try`/`catch` here is enough.

I know that on MacOS if a user navigates to the directory (using Finder) the OS will drop a `.DS_Store` folder there (it stores computed metadata/settings for the directory). We might want to explicitly ignore this directory, or maybe put an extension on the instance name files (i.e., `.json`) so that we can filter for only files that could possibly be relevant (like `if (instanceName.endsWith(".json") ...`).

Reply:
I don't think we _need_ to (invalid files are removed), but I do like the idea of a .json suffix. Originally these files weren't json, but as their complexity grew, I decided it was best.

Reply:
I guess the question was really 'Should we delete files that aren't "ours"?' :thinking:

Reply:
I normally wouldn't be concerned.... but since just using Finder to _read_ directories in turns _writes_ to those directories I'm not sure.